components/heap/esp_heap_debug.c

```diff
@@ -0,0 +1,102 @@
+#include <freertos/FreeRTOS.h>
+#include <freertos/task.h>
+#include <multi_heap.h>
+#include "multi_heap_internal.h"
+//#include "multi_heap_poisoning.h"
+#include "heap_private.h"
+#include "esp_heap_debug.h"
+
+#ifdef CONFIG_HEAP_TASK_TRACKING
+
+// return all possible capabilities (across all priorities) for a given heap
+
+inline static uint32_t get_all_caps(const heap_t *heap)
+{
+    uint32_t all_caps = 0;
+    for (int prio = 0; prio < SOC_MEMORY_TYPE_NO_PRIOS; prio++) {
+        all_caps |= heap->caps[prio];
+    }
+    return all_caps;
+}
+
+// For each task that has allocated memory from the heap, return totals for
+// allocations of each type of heap region (8-bit DRAM, 8-bit D/IRAM, 32-bit
+// IRAM, SPIRAM) into provided array of heap_dump_totals_t structs 'totals'.
+
+// If one or more task handles is provided in array 'tasks', for each block
+// allocated by one of the tasks a heap_dump_block_t struct is written into
+// provided array 'buffer' to give the task, address, size and region type for
+// the block.
+
+size_t esp_heap_debug_dump_totals(heap_dump_totals_t* totals, size_t* ntotal, size_t max,
```

Comment:
This is really useful functionality. I think we can more or less merge the implementation as-is, but I'd like to work on the naming a bit so that it's clearer how to use it (and also try to bring it into line with the existing heap_caps_xxx API).

For a name, how about heap_caps_get_per_task_info()? (With renaming esp_heap_debug.c -> heap_task_info.c, and merging the header into the main esp_heap_caps.h header?)

For heap_dump_totals_t, maybe rename to heap_task_total_t?

I'm not sure how I feel about using the region_info_t enum to split up capabilities, as this isn't how we've treated heap capabilities everywhere else. Instead, would attaching a "uint32_t caps" parameter to this function, so you get aggregate per-task info for a given capabilities mask? In most use cases I think people want to get info on a certain type of memory use (8-bit, or external, or internal) so I think would serve most purposes, and it means this API will match the usage pattern for the other heap metadata APIs. What do you think?

For the (tasks != NULL) behaviour, this seems like it's almost a separate piece of functionality. How would you feel about splitting it into a different function? Something like `heap_caps_get_task_allocations( uint32_t caps, TaskHandle_t *tasks, size_t ntasks, heap_block_info *blocks, size_t nblocks)`?

Reply:
Changing the function and struct names is not a problem.

Regarding the enum for capabilities, I would prefer not to add another 4 bytes to the per-block info struct because the array to store these needs to be quite large and RAM on ESP32 is very tight.  I made my code capable of storing these blocks in 32-bit-only IRAM so that this code intended for debugging heap usage would not place too much burden on the heap itself.

One possibility would be to take an array of pairs of capability masks and values, rather than a single caps value, as an additional parameter and then use the index in that array in place of the enum.  That would allow callers to collect one or more regions as they wish.

Since the parameter list is already longer than one would like, adding more might not be good.  What do you think of the idea of defining a struct to hold all the parameters and having that struct as a single parameter to the function?

I see your point that the per-task totals and the list of individual blocks could be considered separate functionalies.  However, both require a full scan through the heap, which is somewhat expensive and requires locking, and if the heap is active then the results between the totals and the block list could be inconsistent.  Therefore I would prefer to keep this as one function.

Reply:
Partially addressed, pending my comments, in 00fa388.